File: crates/kernel/bpf/xdp_corrupt.c
// XDP byte-level corruption injector. Same principle as xdp_pktloss.c: zero
// randomness in the kernel, every decision about which packet gets touched
// and how comes from userspace. Every Nth packet gets one byte XORed at a
// fixed offset with a fixed mask, both set from userspace at load time, so a
// run is exactly as reproducible as everything else here.
//
// Separate program from xdp_pktloss, XDP only supports one program attached
// per interface per attach mode, this doesn't stack with
// XdpPacketLossInjector on the same interface yet, see the TODO in lib.rs.
#include <linux/bpf.h>
#include <bpf/bpf_helpers.h>

struct bpf_map_def {
    unsigned int type;
    unsigned int key_size;
    unsigned int value_size;
    unsigned int max_entries;
    unsigned int map_flags;
};

// 0 disables corruption entirely, N corrupts every Nth packet.
struct bpf_map_def SEC("maps") corrupt_every_n = {
    .type = BPF_MAP_TYPE_ARRAY,
    .key_size = sizeof(__u32),
    .value_size = sizeof(__u32),
    .max_entries = 1,
};

// Byte offset from the start of the frame to flip. Masked to 0xFFF below
// before use, both so it stays inside any packet XDP will realistically see
// and because the verifier needs a provably bounded value here, it can't
// reason about an arbitrary u32 loaded from a map.
struct bpf_map_def SEC("maps") corrupt_offset = {
    .type = BPF_MAP_TYPE_ARRAY,
    .key_size = sizeof(__u32),
    .value_size = sizeof(__u32),
    .max_entries = 1,
};

// XOR mask applied to the byte at corrupt_offset, only the low 8 bits used.
struct bpf_map_def SEC("maps") corrupt_mask = {
    .type = BPF_MAP_TYPE_ARRAY,
    .key_size = sizeof(__u32),
    .value_size = sizeof(__u32),
    .max_entries = 1,
};

// Own counter, deliberately not shared with xdp_pktloss.c's, these are
// separate programs with independent moduli and shouldn't interfere with
// each other's determinism if both ever end up loaded side by side.
struct bpf_map_def SEC("maps") packet_count = {
    .type = BPF_MAP_TYPE_ARRAY,
    .key_size = sizeof(__u32),
    .value_size = sizeof(__u64),
    .max_entries = 1,
};

SEC("xdp")
int xdp_corrupt(struct xdp_md *ctx)
{
    __u32 key = 0;
    __u32 *n = bpf_map_lookup_elem(&corrupt_every_n, &key);
    if (!n || *n == 0)
        return XDP_PASS;

    __u64 *count = bpf_map_lookup_elem(&packet_count, &key);
    if (!count)
        return XDP_PASS;

    __u64 c = __sync_fetch_and_add(count, 1) + 1;
    if (c % *n != 0)
        return XDP_PASS;

    __u32 *offset = bpf_map_lookup_elem(&corrupt_offset, &key);
    __u32 *mask = bpf_map_lookup_elem(&corrupt_mask, &key);
    if (!offset || !mask)
        return XDP_PASS;

    void *data = (void *)(long)ctx->data;
    void *data_end = (void *)(long)ctx->data_end;

    __u32 off = *offset & 0xFFF; // bound the range so the verifier can prove it
    __u8 *byte = (__u8 *)data + off;

    // verifier needs this exact shape, a computed pointer plus an explicit
    // comparison against data_end right before the access
    if ((void *)(byte + 1) > data_end)
        return XDP_PASS;

    *byte ^= (__u8)(*mask);

    return XDP_PASS;
}

char _license[] SEC("license") = "GPL";

File: crates/kernel/src/xdp_corrupt.rs
use crate::bpf_util::set_u32_map;
use aya::programs::{Xdp, XdpFlags};
use aya::Bpf;
use blackswan_core::{FaultContext, FaultInjector, HarnessError};
use std::convert::TryInto;

const XDP_CORRUPT_OBJ: &[u8] = aya::include_bytes_aligned!(concat!(env!("OUT_DIR"), "/xdp_corrupt.o"));
const PROGRAM_NAME: &str = "xdp_corrupt";
const EVERY_N_MAP: &str = "corrupt_every_n";
const OFFSET_MAP: &str = "corrupt_offset";
const MASK_MAP: &str = "corrupt_mask";

// Deterministic byte-level corruption via XDP. Same shape as
// XdpPacketLossInjector: no randomness in the kernel, arm()/disarm() only
// toggle corrupt_every_n, the hook stays attached once loaded. offset and
// mask are fixed for the injector's lifetime, written once at load time
// rather than on every arm().
//
// Separate program from XdpPacketLossInjector's, can't both be attached to
// the same interface at once yet, see the TODO in lib.rs.
pub struct XdpCorruptionInjector {
    id: String,
    iface: String,
    corrupt_every_n: u32,
    offset: u32,
    mask: u8,
    bpf: Option<Bpf>,
    armed: bool,
}

impl XdpCorruptionInjector {
    pub fn new(id: impl Into<String>, iface: impl Into<String>, corrupt_every_n: u32, offset: u32, mask: u8) -> Self {
        Self {
            id: id.into(),
            iface: iface.into(),
            corrupt_every_n,
            offset,
            mask,
            bpf: None,
            armed: false,
        }
    }

    fn ensure_loaded(&mut self) -> Result<(), HarnessError> {
        if self.bpf.is_some() {
            return Ok(());
        }

        let mut bpf =
            Bpf::load(XDP_CORRUPT_OBJ).map_err(|e| HarnessError::ArmFailed(self.id.clone(), e.to_string()))?;

        {
            let program: &mut Xdp = bpf
                .program_mut(PROGRAM_NAME)
                .ok_or_else(|| {
                    HarnessError::ArmFailed(self.id.clone(), format!("no program named {PROGRAM_NAME} in object"))
                })?
                .try_into()
                .map_err(|e: aya::programs::ProgramError| HarnessError::ArmFailed(self.id.clone(), e.to_string()))?;

            program
                .load()
                .map_err(|e| HarnessError::ArmFailed(self.id.clone(), e.to_string()))?;
            program
                .attach(&self.iface, XdpFlags::default())
                .map_err(|e| HarnessError::ArmFailed(self.id.clone(), e.to_string()))?;
        }

        self.bpf = Some(bpf);

        let bpf = self.bpf.as_mut().expect("just assigned above");
        set_u32_map(bpf, &self.id, OFFSET_MAP, self.offset)?;
        set_u32_map(bpf, &self.id, MASK_MAP, self.mask as u32)?;
        Ok(())
    }
}

impl FaultInjector for XdpCorruptionInjector {
    fn id(&self) -> &str {
        &self.id
    }

    fn arm(&mut self, _ctx: &FaultContext) -> Result<(), HarnessError> {
        self.ensure_loaded()?;
        let bpf = self.bpf.as_mut().expect("just loaded above");
        set_u32_map(bpf, &self.id, EVERY_N_MAP, self.corrupt_every_n)?;
        self.armed = true;
        Ok(())
    }

    fn disarm(&mut self) -> Result<(), HarnessError> {
        let Some(bpf) = self.bpf.as_mut() else {
            self.armed = false;
            return Ok(());
        };
        set_u32_map(bpf, &self.id, EVERY_N_MAP, 0)?;
        self.armed = false;
        Ok(())
    }

    fn is_armed(&self) -> bool {
        self.armed
    }
}
